separatedunit.cpp

```diff
@@ -0,0 +1,1286 @@
+    std::copy(qbListInput.begin(), qbListInput.end(), qbList.begin());
+    std::copy(qbListOutput.begin(), qbListOutput.end(), qbList.begin() + qbListInput.size());
+    carryQbe.cu = qubitLookup[carryIndex].cu;
+    carryQbe.start = qubitLookup[carryIndex].qb;
+    carryQbe.length = 1;
+    qbList[qbList.size() - 1] = carryQbe;
+    OptimizeParallelBitList(qbList);
+
+    EntangleBitList(qbList);
+
+    return coherentUnits[qubitLookup[inputStart].cu]->SbcSuperposeReg8(
+        qubitLookup[inputStart].qb, qubitLookup[outputStart].qb, qubitLookup[carryIndex].qb, values);
+}
+
+/**
+ * Compile an order-preserving list of CoherentUnit bit strings for applying an register-wise operation
+ *
+ * This operation optimizes compiling a list out of qubit pile when bit order is important. We apply register-wise
+ * operations over a pile of arbitrarily entangled and separated qubits. Entangled qubits are stored together in single
+ * CoherentUnit objects, but their mapping to SeparatedUnit bit indices can be generally random. Sometimes, we must
+ * preserve bit order to correctly carry out the operation, whereas sometimes our operation is bitwise parallel and does
+ * not depend on the ordering of bits in the list.
+ */
+void SeparatedUnit::GetOrderedBitList(bitLenInt start, bitLenInt length, std::vector<QbListEntry>& qbList)
+{
+    // Start by getting a list (of sublists) of all the bits we need, with bit sublist length of 1.
+    bitLenInt i, j;
+    QbLookup qbl;
+    QbListEntry qbe;
+    for (i = 0; i < length; i++) {
+        qbl = qubitLookup[start + i];
+        qbe.cu = qbl.cu;
+        qbe.start = qbl.qb;
+        qbe.length = 1;
+        qbList[i] = qbe;
+    }
+
+    // If contiguous sublists in the list we just made are also contiguous in the same coherent unit, we can combine
+    // them to optimize with register-wise gate methods.
+    j = 0;
+    for (i = 0; i < (length - 1); i++) {
+        if ((qbList[j].cu == qbList[j + 1].cu) && ((qbList[j].start + qbList[j].length) == qbList[j + 1].start)) {
+            qbList[j].length += qbList[j + 1].length;
+            qbList.erase(qbList.begin() + j + 1);
+        } else {
+            j++;
+        }
+    }
+}
+
+/// Compile a list of CoherentUnit bit strings for applying a bitwise-parallel operation
+/**
+ * This operation optimizes compiling a list out of qubit pile when bit order is not important. We apply register-wise
+ * operations over a pile of arbitrarily entangled and separated qubits. Entangled qubits are stored together in single
+ * CoherentUnit objects, but their mapping to SeparatedUnit bit indices can be generally random. Sometimes, we must
+ * preserve bit order to correctly carry out the operation, whereas sometimes our operation is bitwise parallel and does
+ * not depend on the ordering of bits in the list.
+ */
+void SeparatedUnit::GetParallelBitList(bitLenInt start, bitLenInt length, std::vector<QbListEntry>& qbList)
```

Comment:
Seems like most of these calls would be simplified with a:
    ```
        CoherentUnit *CohereBitList(std::vector<bitLenInt> bits);
    ```

That would reduce many of the above calls to `CohereBitList(...)->MReg()` or equivalent.

EDIT: Basically EntangleIndices, where it returns a CU.  I'm not sure why many of the calls above start with a GetParallelBitList and then pass that result directly to EntangleIndices.

Reply:
We have both ```GetParallelBitList``` and ```GetOrderedBitList```, depending on whether it's necessary to maintain top-level SU bit order for the application of the particular register-wise gate. ```EntangleBitList``` will accept the result of either method. I'll look at this again and see if I can achieve the same behavior with fewer lines of boilerplate. I'll comment here again with the result.